File: go/99.go
package main

import "math"

type treeState struct {
	prev   *TreeNode
	first  *TreeNode
	second *TreeNode
}

func (s *treeState) traverse(node *TreeNode) {
	if node == nil {
		return
	}

	s.traverse(node.Left)
	if s.prev != nil && s.prev.Val >= node.Val {
		if s.first == nil {
			s.first = s.prev
		}

		s.second = node
	}
	s.prev = node

	s.traverse(node.Right)
}

func recoverTree(root *TreeNode) {
	state := &treeState{
		prev: &TreeNode{Val: math.MinInt64},
	}

	state.traverse(root)

	if state.first != nil && state.second != nil {
		temp := state.first.Val
		state.first.Val = state.second.Val
		state.second.Val = temp
	}
}

File: zig/99.zig
const std = @import("std");

const TreeNode = struct {
    val: i32,
    left: ?*TreeNode = null,
    right: ?*TreeNode = null,
};

const TreeState = struct { prev: ?*TreeNode = null, first: ?*TreeNode = null, second: ?*TreeNode = null };

fn traverse(node: ?*TreeNode, state: *TreeState) void {
    if (node) |n| {
        traverse(n.left, state);
        if (state.prev) |prev| {
            if (prev.val >= n.val) {
                if (state.first == null) {
                    state.first = prev;
                }

                state.second = n;
            }
        }

        state.prev = node;
        traverse(n.right, state);
    }
}

fn solution(root: ?*TreeNode) void {
    if (root == null) {
        return;
    }

    var state = TreeState{};
    traverse(root, &state);

    if (state.first) |first| {
        if (state.second) |second| {
            const temp = first.val;
            first.val = second.val;
            second.val = temp;
        }
    }
}
